File: lib/core/merge.go
package core

import (
	"github.com/TheOpenDictionary/odict/lib/types"
)

// MergeDictionaries merges the entries of two dictionaries.
func MergeDictionaries(dst *types.DictionaryBuffer, srcs ...*types.DictionaryBuffer) (*types.Dictionary, error) {
	parent := dst.Struct()

	for _, src := range srcs {
		entries := src.Struct().Entries

		for _, entry := range entries {
			if _, ok := parent.Entries[entry.Term]; !ok {
				parent.Entries[entry.Term] = entry
			}
		}
	}

	return &parent, nil
}

File: lib/src/core/merge.rs
use crate::Dictionary;

impl Dictionary {
    pub fn merge_multi(&mut self, dictionaries: &[&Dictionary]) {
        for src in dictionaries {
            self.merge(src);
        }
    }

    pub fn merge(&mut self, dictionary: &Dictionary) {
        for (term, entry) in dictionary.entries.iter() {
            if !self.entries.contains_key(term.as_str()) {
                self.entries.insert(term.clone(), entry.clone());
            }
        }
    }
}
